### PR TITLE
Add secure settings validation on startup

### PR DESCRIPTION
**Pull request by rjernst**
Secure settings from the elasticsearch keystore were not yet validated.
This changed improves support in Settings so that secure settings more
seamlessly blend in with normal settings, allowing the existing settings
validation to work. Note that the setting names are still not validated
(yet) when using the elasticsearch-keystore tool.

### CODE REVIEW

**Comment by rjernst** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 1]
```diff
@@ -625,12 +642,28 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
     }
 
     /** Returns the number of settings in this settings object. */
     public int size() {
-        return this.settings.size(); // TODO: account for secure settings
+        return keySet().size();
+    }
+
+    /** Returns the fully qualified setting names contained in this settings object. */
+    public Set<String> keySet() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/Settings.java]: can we document that this makes a real copy?

File: core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java [hunk 2]
```diff
@@ -57,7 +57,7 @@
     private final Setting.Property scope;
     private static final Pattern KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])*[-\\w]+$");
     private static final Pattern GROUP_KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])+$");
-    private static final Pattern AFFIX_KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])+(?:[*][.])+[-\\w]+$");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: can we add a test that catches this bug?

File: core/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 3]
```diff
@@ -625,12 +642,28 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        synchronized (keys) {
+            if (keys.get() == null) {
+                if (secureSettings == null) {
+                    keys.set(settings.keySet());
+                } else {
+                    Stream<String> stream = Stream.concat(settings.keySet().stream(), secureSettings.getSettingNames().stream());
+                    // uniquify, since for legacy reasons the same setting name may exist in both
+                    keys.set(stream.collect(Collectors.toSet()));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/Settings.java]: this should be an immutable set...

**Review by s1monw — APPROVED** [review]
left some suggestions LGTM

File: core/src/main/java/org/elasticsearch/common/settings/Settings.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/Settings.java]: Not sure what you mean by a copy? It lazily constructs it the keyset, but there is only one copy.

File: core/src/main/java/org/elasticsearch/common/settings/Settings.java (hunk 3 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/Settings.java]: Good catch. I always forget collectors do not produce unmodifiable sets. :/

File: core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: I added some tests in 4c84638.

**Pull request merged**